packages/opentelemetry-plugin-xml-http-request/src/xhr.ts

```diff
@@ -0,0 +1,371 @@
+      addSpanNetworkEvent(span, PTN.DOMAIN_LOOKUP_END, resource);
+      addSpanNetworkEvent(span, PTN.CONNECT_START, resource);
+      addSpanNetworkEvent(span, PTN.SECURE_CONNECTION_START, resource);
+      addSpanNetworkEvent(span, PTN.CONNECT_END, resource);
+      addSpanNetworkEvent(span, PTN.REQUEST_START, resource);
+      addSpanNetworkEvent(span, PTN.RESPONSE_START, resource);
+      addSpanNetworkEvent(span, PTN.RESPONSE_END, resource);
+    }
+  }
+
+  /**
+   * Removes the previous information about span.
+   * This might happened when the same xhr is used again.
+   * @param xhr
+   * @private
+   */
+  private _cleanPreviousSpanInformation(xhr: XMLHttpRequestWrapped) {
+    const existingSpanId = xhr.__OT_SPAN_ID;
+    if (existingSpanId) {
+      const callbackToRemoveEvents = this._callbackToRemoveEvents[
+        existingSpanId
+      ];
+      if (callbackToRemoveEvents) {
+        callbackToRemoveEvents();
+      }
+    }
+  }
+
+  /**
+   * Creates a new span when method "open" is called
+   * @param xhr
+   * @param url
+   * @param method
+   * @private
+   */
+  private _createSpan(
+    xhr: XMLHttpRequestWrapped,
+    url: string,
+    method: string
+  ): tracing.Span | undefined {
+    if (isUrlIgnored(url, this._config.ignoreUrls)) {
+      this._logger.debug('ignoring span as url matches ignored url');
+      return;
+    }
+
+    const currentSpan = this._tracer.startSpan(url, {
+      parent: this._tracer.getCurrentSpan() || undefined,
+      attributes: {
+        [AttributeNames.COMPONENT]: this.component,
+        [AttributeNames.HTTP_METHOD]: method,
+        [AttributeNames.HTTP_URL]: url,
+      },
+    }) as tracing.Span;
+
+    currentSpan.addEvent(EventNames.METHOD_OPEN);
+
+    const spanId = currentSpan.context().spanId;
+    this._spans[spanId] = currentSpan;
+
+    this._cleanPreviousSpanInformation(xhr);
+    xhr.__OT_SPAN_ID = spanId;
+
+    return currentSpan;
+  }
+
+  /**
+   * Marks certain [resource]{@link PerformanceResourceTiming} when information
+   * from this is used to add events to span.
+   * This is done to avoid reusing the same resource again for next span
+   * @param resource
+   * @private
+   */
+  private _markResourceAsUsed(resource: PerformanceResourceTiming) {
+    const name = resource.name;
+    if (!this._usedResources[name]) {
+      this._usedResources[name] = [];
+    }
+    this._usedResources[name].push(resource);
+  }
+
+  /**
+   * Patches the method open
+   * @private
+   */
+  protected _patchOpen() {
+    return (original: OpenFunction): OpenFunction => {
+      const plugin = this;
+      return function patchOpen(
+        this: XMLHttpRequest,
+        method: string,
+        url: string,
+        async?: boolean,
+        user?: string | null,
+        pass?: string | null
+      ): OpenFunction | any {
+        plugin._createSpan(this as XMLHttpRequestWrapped, url, method);
+
+        if (async) {
```

Comment:
it is always safer to use `original.apply(this, arguments)` because we can't account for users passing incorrect values or the incorrect number of values. It also removes the need for this check.

Reply:
It is the same here.
You don't have `arguments` word in typescript , so you would still declare array of arguments and define type for it, which would be quite weird anyway. 

Besides that check also the whole code to see more context. The async will always be set to `true` and there are reasons for that [check here](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/open#Parameters). It is also worth to mention that it was ported from open census.

Reply:
You do have the `arguments` keyword in typescript.

It is possible to call open with `async: false` and user not-null. In this case you are changing the arguments the user has called the function with.

We cannot assume that users will use the API properly as documented and it is much easier and safer to directly pass their arguments. If the user incorrectly calls `open` with `async: true` I am suggesting something like this (tests pass with this version):

```typescript
  protected _patchOpen() {
    return (original: Function): OpenFunction => {
      const plugin = this;
      return function patchOpen(
        this: XMLHttpRequest,
        method: string,
        url: string,
        async?: boolean,
        user?: string | null,
        pass?: string | null
      ): ReturnType<OpenFunction> {
        plugin._createSpan(this as XMLHttpRequestWrapped, url, method);
        return original.apply(this, arguments);
        /*
        If you would rather use `call` this can be done easily
        return original.call(this, ...arguments);
        */
      };
    };
  }
```

Regarding the performance difference between apply and call, it may be the case that `call` is faster than `apply`, but in this case the `if` statement would provide more overhead than the difference.

Reply:
still cannot use apply anyway, as I mentioned previously please read the spec, and why it needs to force the async

Reply:
` If true, notification of a completed transaction is provided using event listeners.`

Reply:
What about that means it must be true? If it is false (synchronous), then you just finish the span after calling the original function. If user sets async false in a context that it is not allowed (some browsers only allow sync requests in workers), then 

The way your code is currently, it changes the behavior of user code behind the scenes. If I am a user and I make a synchronous xhr request, I expect it to be made synchronously not transparently rewritten to be async behind the scenes by my telemetry provider.

Reply:
https://github.com/census-instrumentation/opencensus-web/blob/68aa988732e03f2de49ebc0213c3b2525aa5b68b/packages/opencensus-web-instrumentation-zone-peer-dep/src/monkey-patching.ts#L31
and
https://github.com/census-instrumentation/opencensus-web/blob/68aa988732e03f2de49ebc0213c3b2525aa5b68b/packages/opencensus-web-instrumentation-zone-peer-dep/src/monkey-patching.ts#L54

Can you guys @draffensperger @mayurkale22 have your input here too, thx :).

Reply:
and one more  `Synchronous XHR is now in deprecation state.`

Reply:
I understand that it's deprecated, but that doesn't mean users can't or won't use it and the telemetry library shouldn't silently change the behavior of the program.

should we tallk about this in the SIG tomorrow?